### PR TITLE
A11Y: Add keyboard controls to panel menu

### PR DESCRIPTION
**Pull request by kaydelaney**
**What this PR does / why we need it**:
This PR adds keyboard controls to the panel menu as part of the larger accessibility work being done.
In service of this, a new grafana-ui component, PanelMenu, has been added, with a storybook story to assist development and make use of the recently added storybook accessibility add-on.

Making this a draft PR as there are still some aspects I'd like to improve and get feedback on.

See: #35613

### CODE REVIEW

File: packages/grafana-ui/src/components/PanelMenu/PanelMenu.tsx [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+import { useStyles2 } from '../../themes';
+import { PanelMenuListItem } from './PanelMenuItem';
+import { Icon } from '../Icon/Icon';
+import { ClickOutsideWrapper } from '../ClickOutsideWrapper/ClickOutsideWrapper';
+
+interface Props {
+  items: PanelMenuItem[];
+  title: string;
+  buttonClassName?: string;
+  outside?: React.ReactElement;
```
Comment (kaydelaney) [inline packages/grafana-ui/src/components/PanelMenu/PanelMenu.tsx:13]: Not terribly happy with the names "outside" and "inside" here. Need to come up with something more descriptive.

File: packages/grafana-ui/src/components/PanelMenu/PanelMenu.tsx [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+  inside?: React.ReactElement;
+}
+
+const generateMarkup = (item: PanelMenuItem, styles: ReturnType<typeof panelMenuStyles>, items: PanelMenuItem[]) => {
```
Comment (kaydelaney) [inline packages/grafana-ui/src/components/PanelMenu/PanelMenu.tsx:17]: Could probably be moved to a separate file?

File: packages/grafana-ui/src/components/PanelMenu/PanelMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+
+  const onMouseUp = useCallback(() => {
+    setIsMouseDown(false);
+  }, []);
+
+  const onMouseMove = useCallback(() => {
+    if (isMouseDown && !isDragging) {
+      setIsDragging(true);
+    }
+  }, [isMouseDown, isDragging]);
```
Comment (kaydelaney) [inline packages/grafana-ui/src/components/PanelMenu/PanelMenu.tsx:51]: This logic here is all about preventing the menu from coming up when the panel is dragged. Should probably add a comment explaining that, and maybe think of some way to factor it out...

File: packages/grafana-ui/src/components/PanelMenu/PanelMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+      if (isDragging && !menuIsOpen) {
+        return;
+      }
+
+      setMenuOpen(!menuIsOpen);
+    },
+    [menuIsOpen, setMenuOpen, isDragging]
+  );
+
+  const handleKeys = (event: React.KeyboardEvent) => {
```
Comment (kaydelaney) [inline packages/grafana-ui/src/components/PanelMenu/PanelMenu.tsx:66]: I want to refactor all this key-handling logic, and move it to a separate file maybe. Pretty messy and repetitive at the moment.

File: packages/grafana-ui/src/components/PanelMenu/PanelMenu.tsx [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+          ref={menuRef}
+        >
+          {items.map((item) => generateMarkup(item, styles, items))}
+        </ul>
+      </div>
+    </ClickOutsideWrapper>
+  );
+};
+
+const panelMenuStyles = (theme: GrafanaTheme2) => {
```
Comment (kaydelaney) [inline packages/grafana-ui/src/components/PanelMenu/PanelMenu.tsx:188]: Some styles here that should be using theme values probably

File: packages/grafana-ui/src/components/PanelMenu/PanelMenuItem.tsx [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+export const PanelMenuListItem = ({ item, className, children }: Props) => {
+  const styles = useStyles2(panelMenuItemStyles);
+  return (
+    <li className={styles.subMenuItemList} role="none">
+      <button
+        onClick={item.onClick}
+        role="menuitem"
+        tabIndex={-1}
+        className={styles.menuItem}
+        aria-keyshortcuts={item.shortcut}
```
Comment (kaydelaney) [inline packages/grafana-ui/src/components/PanelMenu/PanelMenuItem.tsx:24]: Currently the `shortcut` values passed here don't conform to the spec, so I think they need to be processed before getting passed to `aria-keyshortcuts`.

**Comment by tskarhed** [comment]
Wow, this looks amazing! It also fixes #37436

I'm going to have a go at this with a screen reader and see how it works.

File: packages/grafana-ui/src/components/PanelMenu/PanelMenuItem.tsx [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+          <div className={styles.shortcutContainer}>
+            <Icon name="keyboard" className={styles.shortcutIcon} aria-hidden />
+            <span className={styles.shortcutText} aria-hidden>
+              {item.shortcut}
+            </span>
+          </div>
+        )}
+        {children && <Icon name="angle-right" className={styles.angleRight} aria-hidden />}
+      </button>
+      {children}
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/PanelMenu/PanelMenuItem.tsx:38]: The `More...` menu item causes a bit of confusion. It's probably not that big of a deal, as the submenu items comes directly afterwards.

**Review by tskarhed — COMMENTED** [review]
I don't think we want this component in grafana/ui, unless there is a good example of non-panel applications. At the very least we can have it as an internal.

**Comment by tskarhed** [comment]
Is it possible to incorporate the keyboard navigation in an e2e test?

File: packages/grafana-ui/src/components/PanelMenu/PanelMenuItem.tsx [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+export const PanelMenuListItem = ({ item, className, children }: Props) => {
+  const styles = useStyles2(panelMenuItemStyles);
+  return (
+    <li className={styles.subMenuItemList} role="none">
+      <button
+        onClick={item.onClick}
+        role="menuitem"
+        tabIndex={-1}
```
[the remark this answers is not in the document]
Comment (kaydelaney) [reply packages/grafana-ui/src/components/PanelMenu/PanelMenuItem.tsx:22]: Absolutely right about this, somehow managed to miss it. Working on a fix now!

File: packages/grafana-ui/src/components/PanelMenu/PanelMenu.tsx [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+        const prevSibling = (curMenuItem?.parentElement as HTMLLIElement).previousElementSibling;
+        let prevMenuItem = prevSibling?.firstElementChild as HTMLButtonElement;
+        if (prevSibling && prevSibling.id === 'divider') {
+          prevMenuItem = prevSibling.previousElementSibling?.firstElementChild as HTMLButtonElement;
+        }
+
+        if (!prevMenuItem) {
+          // wrap around to the end if we're at the start of the menu
+          const lastMenuItem = curMenuItem?.parentElement?.parentElement?.lastElementChild
+            ?.firstElementChild as HTMLButtonElement;
```
[the remark this answers is not in the document]
Comment (kaydelaney) [reply packages/grafana-ui/src/components/PanelMenu/PanelMenu.tsx:81]: Not that I'm aware of. Typescript can't know statically what the markup is so it needs hints in order for us to be able to access methods like `focus`, etc.

**Review by torkelo — COMMENTED** [review]
Wish I knew about this plan before, not sure this is a great solution in that it is investing a lot of code into a panel-specific menu component that we should not have. We should extend the Menu component to support nested menus and use that instead. The current panel menu and css classes is in my view legacy and was planed to be replaced by Menu.tsx 

https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/Menu/Menu.tsx 

The goal is even to not need nested menus when we switch to [this](https://github.com/grafana/grafana/issues/31080) design but if we want to address this before the switch to that panel menu we will need nested menu support

**Review by hugohaggmark — COMMENTED** [review]
I guess we can close this PR until the new panel menu design is done?

**Comment by kaydelaney** [comment]
Closed in favor of #38974

**Pull request closed without merging**